// dom-lib/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
var _exportNames = {
  transition: true,
  translateDOMPositionXY: true,
  getVendorPrefixedName: true,
  BrowserSupportCore: true,
  DOMMouseMoveTracker: true,
  WheelHandler: true,
  animation: true
};
exports.animation = exports.WheelHandler = exports.DOMMouseMoveTracker = exports.BrowserSupportCore = exports.getVendorPrefixedName = exports.translateDOMPositionXY = exports.transition = void 0;

var _classNames = require("./classNames");

Object.keys(_classNames).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _classNames[key]) return;
  exports[key] = _classNames[key];
});

var _style = require("./style");

Object.keys(_style).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _style[key]) return;
  exports[key] = _style[key];
});

var _query = require("./query");

Object.keys(_query).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _query[key]) return;
  exports[key] = _query[key];
});

var _events = require("./events");

Object.keys(_events).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _events[key]) return;
  exports[key] = _events[key];
});

var _animation = _interopRequireWildcard(require("./animation"));

exports.animation = _animation;

var _transition = _interopRequireDefault(require("./transition"));

exports.transition = _transition["default"];

var _translateDOMPositionXY = _interopRequireDefault(require("./transition/translateDOMPositionXY"));

exports.translateDOMPositionXY = _translateDOMPositionXY["default"];

var _getVendorPrefixedName = _interopRequireDefault(require("./getVendorPrefixedName"));

exports.getVendorPrefixedName = _getVendorPrefixedName["default"];

var _BrowserSupportCore = _interopRequireDefault(require("./BrowserSupportCore"));

exports.BrowserSupportCore = _BrowserSupportCore["default"];

var _DOMMouseMoveTracker = _interopRequireDefault(require("./DOMMouseMoveTracker"));

exports.DOMMouseMoveTracker = _DOMMouseMoveTracker["default"];

var _WheelHandler = _interopRequireDefault(require("./WheelHandler"));

exports.WheelHandler = _WheelHandler["default"];
