// rsuite-table/lib/utils/useTableDimension.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _react = require("react");

var _domLib = require("dom-lib");

var _constants = require("../constants");

var _debounce = _interopRequireDefault(require("lodash/debounce"));

var _elementResizeEvent = _interopRequireWildcard(require("element-resize-event"));

var _useMount = _interopRequireDefault(require("./useMount"));

var _useUpdateLayoutEffect = _interopRequireDefault(require("./useUpdateLayoutEffect"));

var _isNumberOrTrue = _interopRequireDefault(require("./isNumberOrTrue"));

/**
 * The dimension information of the table,
 * including the height, width, scrollable distance and the coordinates of the scroll handle, etc.
 * @param props
 * @returns
 */
var useTableDimension = function useTableDimension(props) {
  var data = props.data,
      rowHeight = props.rowHeight,
      tableRef = props.tableRef,
      headerWrapperRef = props.headerWrapperRef,
      prefix = props.prefix,
      widthProp = props.width,
      affixHeader = props.affixHeader,
      affixHorizontalScrollbar = props.affixHorizontalScrollbar,
      headerHeight = props.headerHeight,
      height = props.height,
      autoHeight = props.autoHeight,
      children = props.children,
      expandedRowKeys = props.expandedRowKeys,
      onTableWidthChange = props.onTableWidthChange,
      onTableContentWidthChange = props.onTableContentWidthChange,
      onTableContentHeightChange = props.onTableContentHeightChange,
      onTableScroll = props.onTableScroll;
  var contentHeight = (0, _react.useRef)(0);
  var contentWidth = (0, _react.useRef)(0);
  var minScrollY = (0, _react.useRef)(0);
  var scrollY = (0, _react.useRef)(0);
  var scrollX = (0, _react.useRef)(0);
  var minScrollX = (0, _react.useRef)(0);
  var tableWidth = (0, _react.useRef)(widthProp || 0);
  var headerOffset = (0, _react.useRef)(null);
  var tableOffset = (0, _react.useRef)(null);
  var calculateTableContextHeight = (0, _react.useCallback)(function () {
    var prevContentHeight = contentHeight.current;
    var table = tableRef.current;
    var rows = table.querySelectorAll("." + prefix('row')) || [];
    var nextContentHeight = rows.length ? Array.from(rows).map(function (row) {
      return (0, _domLib.getHeight)(row) || rowHeight;
    }).reduce(function (x, y) {
      return x + y;
    }) : 0; // After setting the affixHeader property, the height of the two headers should be subtracted.

    contentHeight.current = nextContentHeight - (affixHeader ? headerHeight * 2 : headerHeight);

    if (!autoHeight) {
      /**
       *  The purpose of subtracting SCROLLBAR_WIDTH is to keep the scroll bar from blocking the content part.
       *  But it will only be calculated when there is a horizontal scroll bar (contentWidth > tableWidth).
       */
      minScrollY.current = -(nextContentHeight - height) - (contentWidth.current > tableWidth.current ? _constants.SCROLLBAR_WIDTH : 0);
    } // If the height of the content area is less than the height of the table, the vertical scroll bar is reset.


    if (nextContentHeight < height) {
      onTableScroll({
        y: 0
      });
    } // If the value of scrollTop is greater than the scrollable range, the vertical scroll bar is reset.
    // When Table is set to virtualized, the logic will be entered every time the wheel event is triggered
    // to avoid resetting the scroll bar after scrolling to the bottom, so add the SCROLLBAR_WIDTH value.


    if (Math.abs(scrollY.current) + height - headerHeight > nextContentHeight + _constants.SCROLLBAR_WIDTH) {
      onTableScroll({
        y: scrollY.current
      });
    }

    if (prevContentHeight !== contentHeight.current) {
      onTableContentHeightChange(prevContentHeight);
    }
  }, [affixHeader, autoHeight, headerHeight, height, onTableScroll, onTableContentHeightChange, prefix, rowHeight, tableRef]);
  var setOffsetByAffix = (0, _react.useCallback)(function () {
    var headerNode = headerWrapperRef === null || headerWrapperRef === void 0 ? void 0 : headerWrapperRef.current;

    if ((0, _isNumberOrTrue["default"])(affixHeader) && headerNode) {
      headerOffset.current = (0, _domLib.getOffset)(headerNode);
    }

    if ((0, _isNumberOrTrue["default"])(affixHorizontalScrollbar) && tableRef !== null && tableRef !== void 0 && tableRef.current) {
      tableOffset.current = (0, _domLib.getOffset)(tableRef === null || tableRef === void 0 ? void 0 : tableRef.current);
    }
  }, [affixHeader, affixHorizontalScrollbar, headerWrapperRef, tableRef]);
  var calculateTableContentWidth = (0, _react.useCallback)(function () {
    var prevWidth = contentWidth.current;
    var table = tableRef === null || tableRef === void 0 ? void 0 : tableRef.current;
    var row = table.querySelector("." + prefix('row') + ":not(.virtualized)");
    var nextContentWidth = row ? (0, _domLib.getWidth)(row) : 0;
    contentWidth.current = nextContentWidth; // The value of SCROLLBAR_WIDTH is subtracted so that the scroll bar does not block the content part.
    // There is no vertical scroll bar after autoHeight.

    minScrollX.current = -(nextContentWidth - tableWidth.current) - (autoHeight ? 0 : _constants.SCROLLBAR_WIDTH);

    if (prevWidth !== contentWidth.current) {
      onTableContentWidthChange(prevWidth);
    }
  }, [autoHeight, onTableContentWidthChange, prefix, tableRef]);
  var calculateTableWidth = (0, _react.useCallback)(function () {
    var prevWidth = tableWidth.current;

    if (tableRef !== null && tableRef !== void 0 && tableRef.current) {
      var nextWidth = (0, _domLib.getWidth)(tableRef === null || tableRef === void 0 ? void 0 : tableRef.current);
      tableWidth.current = nextWidth;
    }

    if (prevWidth !== tableWidth.current) {
      scrollX.current = 0;
      onTableWidthChange === null || onTableWidthChange === void 0 ? void 0 : onTableWidthChange(prevWidth);
    }

    setOffsetByAffix();
  }, [onTableWidthChange, setOffsetByAffix, tableRef]);
  (0, _useMount["default"])(function () {
    calculateTableContextHeight();
    calculateTableContentWidth();
    calculateTableWidth();
    setOffsetByAffix();
    (0, _elementResizeEvent["default"])(tableRef.current, (0, _debounce["default"])(calculateTableWidth, 400));
  });
  (0, _useUpdateLayoutEffect["default"])(function () {
    calculateTableWidth();
    calculateTableContextHeight();
    calculateTableContentWidth();
  }, [data, height, contentHeight, expandedRowKeys, children, calculateTableContextHeight, calculateTableContentWidth]);
  (0, _react.useEffect)(function () {
    return function () {
      if (tableRef.current) {
        // eslint-disable-next-line react-hooks/exhaustive-deps
        (0, _elementResizeEvent.unbind)(tableRef.current);
      }
    }; // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  var setScrollY = (0, _react.useCallback)(function (value) {
    scrollY.current = value;
  }, []);
  var setScrollX = (0, _react.useCallback)(function (value) {
    scrollX.current = value;
  }, []);
  return {
    contentHeight: contentHeight,
    contentWidth: contentWidth,
    minScrollY: minScrollY,
    minScrollX: minScrollX,
    scrollY: scrollY,
    scrollX: scrollX,
    tableWidth: tableWidth,
    headerOffset: headerOffset,
    tableOffset: tableOffset,
    setScrollY: setScrollY,
    setScrollX: setScrollX
  };
};

var _default = useTableDimension;
exports["default"] = _default;
