// dom-lib/lib/animation/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.events = exports.requestAnimationFramePolyfill = exports.nativeRequestAnimationFrame = exports.cancelAnimationFramePolyfill = void 0;

var _query = require("../query");

var _cancelAnimationFramePolyfill = _interopRequireDefault(require("./cancelAnimationFramePolyfill"));

exports.cancelAnimationFramePolyfill = _cancelAnimationFramePolyfill["default"];

var _nativeRequestAnimationFrame = _interopRequireDefault(require("./nativeRequestAnimationFrame"));

exports.nativeRequestAnimationFrame = _nativeRequestAnimationFrame["default"];

var _requestAnimationFramePolyfill = _interopRequireDefault(require("./requestAnimationFramePolyfill"));

exports.requestAnimationFramePolyfill = _requestAnimationFramePolyfill["default"];
var vendorMap = {
  animation: 'animationend',
  OAnimation: 'oAnimationEnd',
  MozAnimation: 'animationend',
  WebkitAnimation: 'webkitAnimationEnd'
};

function getAnimationEvent() {
  if (!_query.canUseDOM) {
    return;
  }

  var tempAnimationEnd;
  var style = document.createElement('div').style;

  for (tempAnimationEnd in vendorMap) {
    if (style[tempAnimationEnd] !== undefined) {
      return vendorMap[tempAnimationEnd];
    }
  }
}

var events = function events() {
  return {
    end: getAnimationEvent()
  };
};

exports.events = events;
